Storage/Storage/CoreData/CoreDataIterativeMigrator.swift

```diff
@@ -59,21 +68,46 @@ final class CoreDataIterativeMigrator {
         var debugMessages = [String]()
 
         do {
-            try steps.forEach { step in
-                let mappingModel = try self.mappingModel(from: step.sourceModel, to: step.targetModel)
-
-                // Migrate the model to the next step
+            // Perform all the migration steps and acquire the last _migrated_ destination URL.
+            let lastTempDestinationURL = try steps.reduce(sourceStoreURL) { currentSourceStoreURL, step in
+                // Log a message
                 let migrationAttemptMessage = makeMigrationAttemptLogMessage(step: step)
                 debugMessages.append(migrationAttemptMessage)
                 DDLogWarn(migrationAttemptMessage)
 
-                try migrateStore(at: sourceStore,
-                                 storeType: storeType,
-                                 fromModel: step.sourceModel,
-                                 toModel: step.targetModel,
-                                 with: mappingModel)
+                // Migrate to temporary URL
+                let tempDestinationURL = try migrate(step: step,
+                                                     sourceStoreURL: currentSourceStoreURL,
+                                                     storeType: storeType)
+
+                // Destroy the `currentSourceStoreURL` if it is a temporary URL since we
+                // will no longer need it.
+                if currentSourceStoreURL != sourceStoreURL {
+                    try persistentStoreCoordinator.destroyPersistentStore(at: currentSourceStoreURL,
+                                                                          ofType: storeType,
+                                                                          options: nil)
+                }
+
+                return tempDestinationURL
             }
 
+            // Now that the migration steps have been performed, replace the store that the
+            // app will use with the _migrated_ store located at the `lastTempDestinationURL`.
+            //
+            // This completes the iterative migration. After this step, the store located
+            // in `sourceStoreURL` should be fully migrated and useable.
+            try persistentStoreCoordinator.replacePersistentStore(at: sourceStoreURL,
+                                                                  destinationOptions: nil,
+                                                                  withPersistentStoreFrom: lastTempDestinationURL,
+                                                                  sourceOptions: nil,
+                                                                  ofType: storeType)
+
+            // Final clean-up. Destroy the store at `lastTempDestinationURL` since it should have
+            // been copied over to `sourceStoreURL` during `replacePersistentStore` (above).
+            try persistentStoreCoordinator.destroyPersistentStore(at: lastTempDestinationURL,
```

Comment:
Good thought-provoking question! 

It's possible if the [steps](https://github.com/woocommerce/woocommerce-ios/blob/47bb9ed7243f0c261cd7d61b1d81ea81f81f6ec0/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L63) turn out to be empty. But we have the `guard` statement:

https://github.com/woocommerce/woocommerce-ios/blob/47bb9ed7243f0c261cd7d61b1d81ea81f81f6ec0/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L64

The following tests also fail if the `steps` array returns empty because of an incorrect logic:

https://github.com/woocommerce/woocommerce-ios/blob/47bb9ed7243f0c261cd7d61b1d81ea81f81f6ec0/Storage/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift#L165

https://github.com/woocommerce/woocommerce-ios/blob/47bb9ed7243f0c261cd7d61b1d81ea81f81f6ec0/Storage/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift#L206

So I think we're fairly protected. I'll just leave it as is to keep the code shorter and hopefully easier to understand.